### PR TITLE
Display Window or Global on Window class reps

### PR DESCRIPTION
**Pull request by Zacqary**
Associated Issue: #884 

### Summary of Changes

* Window rep will display either `"Window"` or `"Global"` depending on whether its preview has a truthy `isGlobal` property or not
* Top-level Window will override its preview to `{ kind: "Window", isGlobal: true }`

### Notes

This feels like a hack to me, and I'm submitting this PR more to start a discussion than because I'm confident in its ability to fix the issue. I don't understand the code path that renders the reps very well, but this is what I was able to figure out.

The problem that I found is that the top-level Window — the one we want to say `Global` — has a `preview` with `{ kind: "ObjectWithURL" }` rather than `{ kind: "Window" }`. But, apparently, the `view` property in `e` is correctly marked as `{ kind: "Window" }`. I have no idea why it is, or where the `view` property is even coming from, because it's not present in `getBindingVariables`. It seems to be added at some step before the final render, but I wasn't able to track that down.

So that's why I'm not super confident in this fix. It operates under the assumption that if a scope of type `"object"` has a class of `"Window"`, then it must be the global `Window`. I don't know if this is always going to be true. It also doesn't fix the underlying problem of why the global `Window` is being generated with the wrong kind of preview in the first place, which I also wasn't able to track down.

### Testing

* [x] passes `npm test`
* [x] passes `npm run lint`

### Screenshots
screen shot 2016-10-10 at 2 56 46 pm

### CODE REVIEW

**Comment by jasonLaster** [comment]
Thanks @Zacqary! I think you're on the right track, but I'm going to bump this review to @janodvarko, @linclark, who have worked more on Reps than I have.

File: public/js/components/Scopes.js [hunk 1]
```diff
@@ -120,10 +120,18 @@ function getScopes(pauseInfo, selectedFrame) {
...
       }
     } else if (type === "object") {
+      const value = scope.object;
+      // Override preview for Window object
+      if (value.class === "Window") {
+        value.preview = {
+          kind: "Window",
+          isGlobal: true
+        };
+      }
```
Comment (jasonLaster) [inline public/js/components/Scopes.js]: hmm, interesting - i'm not sure if we have mutated the rep value's before. Curious to see what others think.

File: public/js/lib/devtools/client/shared/components/reps/window.js [hunk 2]
```diff
@@ -37,8 +37,8 @@ define(function (require, exports, module) {
       return "";
     },
 
-    getLocation: function (grip) {
-      return getURLDisplayString(grip.preview.url);
+    getDisplayValue: function (grip) {
+      return grip.preview.isGlobal ? "Global" : "Window";
```
Comment (jasonLaster) [inline public/js/lib/devtools/client/shared/components/reps/window.js]: Thanks @janodvarko. 

I think updating the *tiny* mode makes a lot of sense. We're already forcing our grip's to be [tiny](https://github.com/devtools-html/debugger.html/blob/master/public/js/components/ObjectInspector.js#L136).

The motivation behind formatting the window object with `Window` or `Global` is in part space constraints and in part it's clearer, in the same way that labeling Arguments as opposed to referring to it as Object. The discussion is here #884.

File: public/js/components/Scopes.js (hunk 1 above)
Comment (jlongster) [inline public/js/components/Scopes.js]: Yeah, I think we should avoid it. We want to be careful about mutations because they cause surprising effects down the road. It might be good to ask Honza about this.

Comment (jlongster) [inline public/js/components/Scopes.js]: Oh, just noticed that he commented below. Sounds like we have a different approach in mind now?

File: public/js/lib/devtools/client/shared/components/reps/window.js (hunk 2 above)
Comment (Zacqary) [inline public/js/lib/devtools/client/shared/components/reps/window.js]: Just updated the Window rep to use `tiny` mode, and handled attaching the `isGlobal` property by cloning the rep value instead of mutating it.

I didn't realize `kind = ObjectWithUrl` was getting routed correctly to the `Window` rep component, so I also removed the kind change.

I'm still not sure I entirely understand the implications of the `else if (type === "object")` block where the `isGlobal` prop gets added, but it seems to work.

File: public/js/components/Scopes.js (hunk 1 above)
Comment (janodvarko) [inline public/js/components/Scopes.js]: Yes, different approach suggested. And I agree, we should avoid rep value mutations.

**Comment by jasonLaster** [comment]
Thanks @Zacqary! This looks good to me.

File: public/js/components/Scopes.js [hunk 3]
```diff
@@ -120,10 +120,16 @@ function getScopes(pauseInfo, selectedFrame) {
...
+      // If this is the global window scope, mark it as such so that it will
+      // preview Window: Global instead of Window: Window
+      if (value.class === "Window") {
+        value = Object.assign({}, scope.object, { isGlobal: true });
+      }
       scopes.push({
         name: scope.object.class,
         path: key,
-        contents: { value: scope.object }
+        contents: { value }
```
Comment (jasonLaster) [inline public/js/components/Scopes.js]: I like the way that this updates the grip without mutating the original source.

File: public/js/lib/devtools/client/shared/components/reps/window.js [hunk 4]
```diff
@@ -41,14 +42,22 @@ define(function (require, exports, module) {
       return getURLDisplayString(grip.preview.url);
     },
 
+    getDisplayValue: function (grip) {
+      if (this.props.mode === "tiny") {
+        return grip.isGlobal ? "Global" : "Window";
+      } else {
+        return this.getLocation(grip);
+      }
+    },
```
Comment (jasonLaster) [inline public/js/lib/devtools/client/shared/components/reps/window.js]: I like that this doesn't change the original behavior, except for the tiny mode.

File: public/js/lib/devtools/client/shared/components/reps/window.js (hunk 2 above)
Comment (jasonLaster) [inline public/js/lib/devtools/client/shared/components/reps/window.js]: @janodvarko mind taking another look. Once this lands, I'd be happy to make an m-c patch and perhaps add some m-c tests.

File: public/js/components/Scopes.js (hunk 3 above)
Comment (janodvarko) [inline public/js/components/Scopes.js]: Yes, looks correct to me to.

File: public/js/lib/devtools/client/shared/components/reps/window.js (hunk 4 above)
Comment (janodvarko) [inline public/js/lib/devtools/client/shared/components/reps/window.js]: Nice, that's exactly what 'modes' are for.

**Comment by jasonLaster** [comment]
Thanks @Zacqary! That was not an *easy* one :)

**Comment by Zacqary** [comment]
No problem. Yeah, definitely a challenge but I've gotten pretty good at
grepping through the breadcrumbs of React codebases.

The one thing I still wasn't able to figure out was where all that
additional stuff, like the view prop, was getting added to the event object
before rendering. The props chain gets very hard to follow somewhere in the
middle of it all, with all the renders of component classes that get
computed from lists.

Also, are there any docs on what exactly all the terms like "rep" and
"grip" mean?

**Comment by jasonLaster** [comment]
> props chain very hard to follow somewhere in the middle of it all, with all the renders of component classes that get computed from lists.

Yep and you're in a very hard spot with the Tree + Rep... The sources tree is also confusing for other reasons, but this is as complicated as it gets :)

> Also, are there any docs on what exactly all the terms like "rep" and "grip" mean?

:) I don't think so... @janodvarko would likely know. This would be tremendously helpful if we try to unify Firefox and Chrome's definition of remote objects. https://bugzilla.mozilla.org/show_bug.cgi?id=1307877

**Comment by janodvarko** [comment]
> Also, are there any docs on what exactly all the terms like "rep" and "grip" mean?

Grip comes from RDP protocol:
https://wiki.mozilla.org/Remote_Debugging_Protocol#Grips

Rep stands for representation of data.

Honza
